### PR TITLE
Add client methods to support Actions artifact APIs.

### PR DESCRIPTION
**Pull request by lerebear**
Partially addresses https://github.com/octokit/octokit.rb/issues/1216.

This adds support for the Actions artifact APIs documented at https://docs.github.com/en/rest/actions/artifacts.

## Help needed please 🙏🏿 

- [x] As a first-time contributor, I need a maintainer to approve my pending Actions workflows on this PR please. 
- [x] It isn't clear to me how to record a new cassette for these tests. I've tried hardcoding the [default record_mode](https://github.com/lerebear/octokit.rb/blob/52e099abe3b355d44d921ea10bc7979a47b1df59/spec/helper.rb#L144) to `:once` and running tests with either `script/test` or `bundle exec rspec spec/octokit/client/actions_artifacts_spec.rb`, both to no avail. The lack of cassettes for these new tests will likely cause them to fail once the pending Actions workflows are authorized, so any guidance here would be much appreciated.
- [x] I'm not sure how to satisfy the documentation requirement from the contribution guide. I've run `bundle exec rake doc:yard` locally, but I'm not sure how to interpret its output. Please let me know if there's anything more to do here.

### CODE REVIEW

**Comment by Brend-Smits** [comment]
> This looks great to me from a code perspective ✨ I don't think there is anything you need to worry about from a documentation perspective as this all looks good to me. I'll make a note to revisit `CONTRIBUTING.md`.
> 
> I'm not sure how to get VCR working - I've had similar issues. But @Brend-Smits managed to get round this with his recent PR, #1478, so I am wondering if he might be able to help?

VCR automatically recorded cassettes for me. It only does so once though.
You might need to set environment variables to enable record mode. See the helper method here: https://github.com/octokit/octokit.rb/blob/main/spec/helper.rb#L130-L148

**Comment by lerebear** [comment]
> It isn't clear to me how to record a new cassette for these tests.

I figured this out with a some help from @juanmrad 🙌🏿  (another recent contributor to this project).

In my case, I had misunderstood the interaction between Webmock and VCR. Specifically, my tests originally used Webmock stubs like so:

https://github.com/octokit/octokit.rb/blob/52e099abe3b355d44d921ea10bc7979a47b1df59/spec/octokit/client/actions_artifacts_spec.rb#L27

When VCR encounters that request, it recognizes it as "externally stubbed" and ignores it.

Once I removed those stubs, then VCR began attempting to record cassettes for new requests as expected. From there, I did indeed have to set the environment variables described in the [docs on running and writing new tests](https://github.com/octokit/octokit.rb#running-and-writing-new-tests) that were relevant to my tests. Those were `OCTOKIT_TEST_GITHUB_LOGIN`, `OCTOKIT_TEST_GITHUB_TOKEN` and `OCTOKIT_TEST_GITHUB_REPOSITORY`. To avoid needing to set unnecessary environment variables, I took two additional steps:

1. I ran just my newly added test suite with `bundle exec rspec spec/octokit/client/actions_artifacts_spec.rb` rather than `script/test`.
2. I commented out the following portion of the test setup, which assumes that the tests need a real organization-owned repository (which mine do not): https://github.com/octokit/octokit.rb/blob/15be7251e795044f2bb9537db7c58c95fdd4cbec/spec/helper.rb#L118-L123

With those changes in place I was able to record and replay cassettes as expected.

One last note is that the [VCR debug logging](https://relishapp.com/vcr/vcr/v/2-1-0/docs/configuration/debug-logging) config was pretty helpful in making the above discoveries; I enabled that along the way too.

**Comment by lerebear** [comment]
@timrogers would you mind providing another review of this please? I think my pending workflows need to be approved again too.

I also wonder if you can explain the remainder of the release process? Specifically I'm curious as to:

- Whether or not I should merge this pull request myself once its approved
- When this feature might be released in a new version of this library
- If features like this are ever backported to older versions of Octokit (specifically 4.x)

Thanks!

**Comment by timrogers** [comment]
> @timrogers would you mind providing another review of this please? I think my pending workflows need to be approved again too.

Sure!

> I also wonder if you can explain the remainder of the release process? Specifically I'm curious as to:
> 
> * Whether or not I should merge this pull request myself once its approved
> * When this feature might be released in a new version of this library
> * If features like this are ever backported to older versions of Octokit (specifically 4.x)

Of course:

1. One of the maintainers will take care of that once a PR has been approved.
2. We generally release a new version as soon as new features are merged.
3. We have never done that historically, but we could do if there was clear value to it.

File: lib/octokit/client/actions_artifacts.rb [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    # Methods for the Actions Artifacts API
+    #
+    # @see https://developer.github.com/v3/actions/artifacts
+    module ActionsArtifacts
+      # List all artifacts for a repository
+      #
+      # @param repo [Integer, String, Repository, Hash] A GitHub repository
+      #
+      # @return [Sawyer::Resource] the total count and an array of artifacts
+      # @see https://developer.github.com/v3/actions/artifacts#list-artifacts-for-a-repository
```
Comment (timrogers) [inline lib/octokit/client/actions_artifacts.rb:14]:
```suggestion
      # @see https://docs.github.com/en/rest/actions/artifacts#list-artifacts-for-a-repository
```

File: lib/octokit/client/actions_artifacts.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        paginate "#{Repository.path repo}/actions/artifacts", options
+      end
+
+      # List all artifacts for a workflow run
+      #
+      # @param repo [Integer, String, Repository, Hash] A GitHub repository
+      # @param workflow_run_id [Integer] Id of a workflow run
+      #
+      # @return [Sawyer::Resource] the total count and an array of artifacts
+      # @see https://developer.github.com/v3/actions/artifacts#list-workflow-run-artifacts
```
Comment (timrogers) [inline lib/octokit/client/actions_artifacts.rb:25]:
```suggestion
      # @see https://docs.github.com/en/rest/actions/artifacts#list-workflow-run-artifacts
```

File: lib/octokit/client/actions_artifacts.rb [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        paginate "#{Repository.path repo}/actions/runs/#{workflow_run_id}/artifacts", options
+      end
+
+      # Get an artifact
+      #
+      # @param repo [Integer, String, Repository, Hash] A GitHub repository
+      # @param id [Integer] Id of an artifact
+      #
+      # @return [Sawyer::Resource] Artifact information
+      # @see https://developer.github.com/v3/actions/artifacts#get-an-artifact
```
Comment (timrogers) [inline lib/octokit/client/actions_artifacts.rb:36]:
```suggestion
      # @see https://docs.github.com/en/rest/actions/artifacts#get-an-artifact
```

File: lib/octokit/client/actions_artifacts.rb [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        get "#{Repository.path repo}/actions/artifacts/#{id}", options
+      end
+
+      # Get a download URL for an artifact
+      #
+      # @param repo [Integer, String, Repository, Hash] A GitHub repository
+      # @param id [Integer] Id of an artifact
+      #
+      # @return [String] URL to the .zip archive of the artifact
+      # @see https://developer.github.com/v3/actions/workflow-runs/#download-workflow-run-logs
```
Comment (timrogers) [inline lib/octokit/client/actions_artifacts.rb:47]:
```suggestion
      # @see https://docs.github.com/en/rest/actions/workflow-runs#download-workflow-run-logs
```

File: lib/octokit/client/actions_artifacts.rb [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        response.headers['Location']
+      end
+
+      # Delete an artifact
+      #
+      # @param repo [Integer, String, Repository, Hash] A GitHub repository
+      # @param id [Integer] Id of an artifact
+      #
+      # @return [Boolean] Return true if the artifact was successfully deleted
+      # @see https://developer.github.com/v3/actions/artifacts#delete-an-artifact
```
Comment (timrogers) [inline lib/octokit/client/actions_artifacts.rb:61]:
```suggestion
      # @see https://docs.github.com/en/rest/actions/artifacts#delete-an-artifact
```

File: spec/octokit/client/actions_artifacts_spec.rb [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+      expect(url).to be_kind_of String
+    end
+  end
+
+  describe '.delete_artifact' do
+    it 'deletes the artifact' do
+      status = @client.delete_artifact(@test_repo, @artifact_id)
+
+      assert_requested :delete, github_url("repos/#{@test_repo}/actions/artifacts/#{@artifact_id}")
+      expect(status).to be(true).or be(false)
```
Comment (timrogers) [inline spec/octokit/client/actions_artifacts_spec.rb:58]: Given that we have a mock recorded with VCR, we should be able to be certain about the return value we expect.

```suggestion
      expect(status).to be(true)
```

**Comment by timrogers** [comment]
Released to RubyGems in v5.6.0.

**Comment by timrogers** [comment]
> > It isn't clear to me how to record a new cassette for these tests.
> 
> I figured this out with a some help from @juanmrad 🙌🏿 (another recent contributor to this project).
> 
> In my case, I had misunderstood the interaction between Webmock and VCR. Specifically, my tests originally used Webmock stubs like so:
> 
> https://github.com/octokit/octokit.rb/blob/52e099abe3b355d44d921ea10bc7979a47b1df59/spec/octokit/client/actions_artifacts_spec.rb#L27
> 
> When VCR encounters that request, it recognizes it as "externally stubbed" and ignores it.
> 
> Once I removed those stubs, then VCR began attempting to record cassettes for new requests as expected. From there, I did indeed have to set the environment variables described in the [docs on running and writing new tests](https://github.com/octokit/octokit.rb#running-and-writing-new-tests) that were relevant to my tests. Those were `OCTOKIT_TEST_GITHUB_LOGIN`, `OCTOKIT_TEST_GITHUB_TOKEN` and `OCTOKIT_TEST_GITHUB_REPOSITORY`. To avoid needing to set unnecessary environment variables, I took two additional steps:
> 
> 1. I ran just my newly added test suite with `bundle exec rspec spec/octokit/client/actions_artifacts_spec.rb` rather than `script/test`.
> 2. I commented out the following portion of the test setup, which assumes that the tests need a real organization-owned repository (which mine do not): https://github.com/octokit/octokit.rb/blob/15be7251e795044f2bb9537db7c58c95fdd4cbec/spec/helper.rb#L118-L123
> 
> With those changes in place I was able to record and replay cassettes as expected.
> 
> One last note is that the [VCR debug logging](https://relishapp.com/vcr/vcr/v/2-1-0/docs/configuration/debug-logging) config was pretty helpful in making the above discoveries; I enabled that along the way too.

I'd very much welcome a contribution to the docs on this ✨

**Pull request merged**